### PR TITLE
Add ctor to JsonSerializerOptions that takes serializer defaults

### LINKED ISSUE #34626: Don't forget about JsonSerializerOptions.CreateForWeb()
This is a leftover from https://github.com/dotnet/runtime/issues/32937, we didn't add the default options that are meant to sync the options between `System.Net.Http.Json` and ASP.NET.

As per the [spec](https://github.com/dotnet/designs/blob/master/accepted/2020/json-http-extensions/json-http-extentions.md#default-serialization-options), the next API should be added as well.

```cs
namespace System.Text.Json
{
    public partial class JsonSerializerOptions
    {
        public static JsonSerializerOptions CreateForWeb();
    }
}
```

Above method should return an options instance as the following:
```cs
new JsonSerializerOptions 
{ 
    PropertyNameCaseInsensitive = true, 
    PropertyNamingPolicy = JsonNamingPolicy.CamelCase 
}
```

### PR DESCRIPTION
**Pull request by NikiforovAll**
Fixes #34626

I decided to take over because of no activity of the current assignee. Let me know if I'm moving in the right direction.

### CODE REVIEW

**Comment by NikiforovAll** [comment]
skipped [docs/coding-guidelines/updating-ref-source.md](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/updating-ref-source.md): ref-source was updated manually

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs
Comment (stephentoub) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:228]: The `: byte` should be removed, along with the casts to `byte` below.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerDefaults.cs
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerDefaults.cs:8]:
```suggestion
    /// Signifies what default options are used by <see cref="JsonSerializerOptions"/>.
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerDefaults.cs
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerDefaults.cs:14]:
```suggestion
        /// <summary>
        /// Specifies that general-purpose values should be used.  These are the same settings applied if a <see cref="JsonSerializerDefaults"/> isn't specified.
        /// </summary>
        /// <remarks>
        /// This option implies that property names are treated as case-sensitive and that "PascalCase" name formatting should be employed.
        /// </remarks>
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerDefaults.cs
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerDefaults.cs:18]: /// <summary>
        /// Specifies that values should be used more appropriate to web-based scenarios.
        /// </summary>
        /// <remarks>
        /// This option implies that property names are treated as case-insensitive and that "camelCase" name formatting should be employed.
        /// </remarks>

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:89]:
```suggestion
        /// Constructs a new <see cref="JsonSerializerOptions"/> instance with a predefined set of options determined by the specified <see cref="JsonSerializerDefaults"/>.
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:98]:
```suggestion
            }
            else if (defaults != JsonSerializerDefaults.General)
            {
                throw new ArgumentOutOfRangeException(nameof(defaults));
            }
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:101]: Can you please add tests as well?

**Review by stephentoub — COMMENTED** [review]
Thanks for working on this.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs
Comment (NikiforovAll) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:101]: Sure, thanks for the review.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs
Comment (NikiforovAll) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:98]: This makes code more robust. But, it is not possible to test it at the moment. WDYT?

Sweet 100% code coverage. 🤔

Comment (stephentoub) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:98]: Why isn't it possible to test it?

**Comment by layomia** [comment]
Pinging ASP.NET folks and @terrajobst as there was some discussion about whether `options.Encoder` should be set to `JavaScriptEncoder.UnsafeRelaxedJsonEscaping` when using `Web` defaults - https://github.com/dotnet/aspnetcore/pull/21016/files#r411571846.

@pranavkm previously mentioned that `UnsafeRelaxedJsonEscaping` is only safe if the JSON isn't interspersed with markup, and that `Web` defaults might not convey that this is meant for REST/API scenarios. I just wanted to be sure about the conclusion here before merging this.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs
Comment (rynowak) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:97]: @layomia - asked me to drop by and confirm whether this matches ASP.NET's expectations.  The relevant code is [here](https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.Core/src/JsonOptions.cs#L29) for anyone who wants to compare.

Some items worth of discussion:

- `MaxDepth`
- *relaxed encoder*

For MaxDepth - we did not scientifically compute the number for maxdepth. We have just always had a value of 32 - and yours is 64. I don't foresee a problem with us using 64. So I think we can agree to use your default of 64.

For the *encoder* - (this is the fun one) we do some pretty subtle stuff. Our main options class doesn't specify an encoder, and we special case it in two places:

- For writing output as JSON response from the server [we opt in the *relaxed* encoder](https://github.com/dotnet/aspnetcore/blob/4191eb5c7629ddbf88168ac45cffb7bea3b7e956/src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs)
- For writing [JSON in HTML](https://github.com/dotnet/aspnetcore/blob/4191eb5c7629ddbf88168ac45cffb7bea3b7e956/src/Mvc/Mvc.ViewFeatures/src/Rendering/SystemTextJsonHelper.cs#L29) we force the default (strict) encoder 

If we want to be technical - JSON as a document/request/response with the content type `application/json` is always unicode. The JSON specs all agree, that JSON is all unicode all of the time. We use the relaxed encoder when sending JSON over the wire as its own self-contained payload because all consumers should expect it to the unicode.

We *don't* use the relaxed encoder when JSON is being embedded inside an HTML response, because that's browser-land, and it's much more of a wierd place.

I think regardless of which encoder is used as the default we'll keep our behavior the same. I won't try to cosplay as a security expert. My assumption is that the common task using the "web" settings is to read and write JSON documents for communication with a REST API, so my intuition is that using the relaxed encoding is right.

/cc @GrabYourPitchforks @pranavkm

Comment (NikiforovAll) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:97]: > My assumption is that the common task using the "web" settings is to read and write JSON documents for communication with a REST API, so my intuition is that using the relaxed encoding is right.

I agree with you on that. Although, just `Web` option would mislead code clients to potential vulnerability issues. We could provide a more verbose name that communicates usage and intentions. Something like `DO_NO_USE_IT_IN_HTML__Web` JK

**Comment by NikiforovAll** [comment]
@layomia. Could you please clarify the final decision regarding the Encoder?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:97]: I discussed with @terrajobst - since specifying the unsafe relaxed encoder isn't recommended in some scenarios, we shouldn't specify it in a generic "web" default. Developers should specify it atop the default, depending on the scenario, as is done in an example @rynowak provided above. So, the current implementation works fine.

We don't want to have various "web" default variants (WebRest, WebHtml etc) as this would likely introduce the overhead of figuring out what the differences are, particularly if the set of defaults grows.

We'll leave max depth (effective) as 64, since we don't foresee issues with ASP.NET switching to this number.

File: src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs:556]:
```suggestion
            Assert.Same(JsonNamingPolicy.CamelCase, options.PropertyNamingPolicy);
```

- `Assert.NotNull` isn't needed here
- The expected value should be the first arg
- `Assert.Same` is used when we expect the same object reference

File: src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/OptionsTests.cs:563]: Let's test an input value of `2` as well.

**Review by layomia — APPROVED** [review]
Just a couple comments, otherwise LGTM. We can merge when addressed & conflicts are fixed.

**Comment by NikiforovAll** [comment]
@layomia I assume you want me to add/update [.NET API Reference Docs](https://github.com/dotnet/dotnet-api-docs). Also, I would like to add a little section to [docs/standard/serialization/system-text-json-how-to.md](https://github.com/dotnet/docs/blob/master/docs/standard/serialization/system-text-json-how-to.md), something like _"Use predefined JsonSerializerOptions"_, but I'm not sure if it is valuable enough.

**Comment by NikiforovAll** [comment]
@layomia
Unfortunately, I've found guidance on how to update API Reference Docs. Could you please assist? 🙂

**Comment by layomia** [comment]
@NikiforovAll, I didn't intend for you to update the docs. We'll handle that e.g. with https://github.com/dotnet/dotnet-api-docs/pull/4284. I was assigning authors to PRs as part of triage. Thanks for reaching out, and sorry for the noise.

**Pull request merged**
